contracts/access/manager/AccessManager.sol

```diff
@@ -0,0 +1,564 @@
+        address caller = _msgSender();
+        bytes4 selector = bytes4(data[0:4]);
+
+        // Fetch restriction to that apply to the caller on the targeted function
+        (bool allowed, uint32 setback) = canCall(caller, target, selector);
+
+        // If caller is not authorised, revert
+        if (!allowed && setback == 0) {
+            revert AccessManagerUnauthorizedCall(caller, target, selector);
+        }
+
+        // If caller is authorised, schedule operation
+        bytes32 operationId = _hashOperation(caller, target, data);
+        if (_schedules[operationId] != 0) {
+            revert AccessManagerAlreadyScheduled(operationId);
+        }
+        _schedules[operationId] = Time.timestamp();
+
+        emit Scheduled(operationId, caller, target, data);
+        return operationId;
+    }
+
+    /**
+     * @dev Execute a function that is delay restricted, provided it was properly scheduled beforeend, or the
+     * execution delay is 0.
+     *
+     * Emits a {Executed} event if the call was scheduled. Unscheduled call (with no delay) do not emit that event.
+     */
+    function relay(address target, bytes calldata data) public payable virtual {
+        address caller = _msgSender();
+        bytes4 selector = bytes4(data[0:4]);
+
+        // Fetch restriction to that apply to the caller on the targeted function
+        (bool allowed, uint32 setback) = canCall(caller, target, selector);
+
+        // If caller is not authorised, revert
+        if (!allowed && setback == 0) {
+            revert AccessManagerUnauthorizedCall(caller, target, selector);
+        }
+
+        // If caller is authorised, check operation was scheduled early enough
+        bytes32 operationId = _hashOperation(caller, target, data);
+        uint48 timepoint = _schedules[operationId];
+        if (setback != 0) {
+            if (timepoint == 0) {
+                revert AccessManagerNotScheduled(operationId);
+            }
+            if (timepoint + setback > Time.timestamp()) {
+                revert AccessManagerNotReady(operationId);
+            }
+        }
+        if (timepoint != 0) {
+            delete _schedules[operationId];
+            emit Executed(operationId);
+        }
+
+        // Mark the target and selector as authorised
+        bytes32 relayIdentifierBefore = _relayIdentifier;
+        _relayIdentifier = keccak256(abi.encodePacked(target, selector));
+
+        // Perform call
+        Address.functionCallWithValue(target, data, msg.value);
+
+        // Reset relay identifier
+        _relayIdentifier = relayIdentifierBefore;
+    }
+
+    /**
+     * @dev Cancel a scheduled (delayed) operation.
+     *
+     * Requirements:
+     *
+     * - the caller must be the proposer, or a guardian of the targeted function
+     *
+     * Emits a {Canceled} event.
+     */
+    function cancel(address caller, address target, bytes calldata data) public virtual {
```

Comment:
One possible issue we have is that a guardian could cancel an operation that is part of a governance proposal. It seems like that would cause the whole batch to revert, we don't have many other options.